File: tests/fixtures/sample_project/users/crud.py
"""User model + CRUD operations."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone

from ..db.connection import get_connection


@dataclass
class User:
    """Canonical user record."""

    id: int
    email: str
    password_hash: str
    created_at: datetime
    last_login: datetime | None = None
    is_active: bool = True


def create_user(email: str, password_hash: str) -> User:
    """Insert a new user and return the created record."""
    conn = get_connection()
    cursor = conn.execute(
        "INSERT INTO users (email, password_hash, created_at, is_active) "
        "VALUES (?, ?, ?, ?) RETURNING id",
        (email, password_hash, datetime.now(timezone.utc), True),
    )
    new_id = cursor.fetchone()[0]
    return User(
        id=new_id,
        email=email,
        password_hash=password_hash,
        created_at=datetime.now(timezone.utc),
    )


def get_user_by_email(email: str) -> dict | None:
    """Fetch a user by email, returning a raw dict (for auth hot path speed)."""
    conn = get_connection()
    row = conn.execute(
        "SELECT id, email, password_hash, created_at, last_login, is_active "
        "FROM users WHERE email = ? AND is_active = 1",
        (email,),
    ).fetchone()
    if row is None:
        return None
    return {
        "id": row[0],
        "email": row[1],
        "password_hash": row[2],
        "created_at": row[3],
        "last_login": row[4],
        "is_active": bool(row[5]),
    }


def get_user(user_id: int) -> User | None:
    """Fetch a user by primary key."""
    conn = get_connection()
    row = conn.execute(
        "SELECT id, email, password_hash, created_at, last_login, is_active "
        "FROM users WHERE id = ?",
        (user_id,),
    ).fetchone()
    if row is None:
        return None
    return User(
        id=row[0],
        email=row[1],
        password_hash=row[2],
        created_at=row[3],
        last_login=row[4],
        is_active=bool(row[5]),
    )


def update_last_login(user_id: int) -> None:
    """Stamp the user's last_login timestamp."""
    conn = get_connection()
    conn.execute(
        "UPDATE users SET last_login = ? WHERE id = ?",
        (datetime.now(timezone.utc), user_id),
    )


def deactivate_user(user_id: int) -> None:
    """Soft-delete a user by flipping is_active off."""
    conn = get_connection()
    conn.execute("UPDATE users SET is_active = 0 WHERE id = ?", (user_id,))

File: tests/fixtures/sample_project_ts/src/users/crud.ts
// User model + CRUD operations.

import { getConnection } from "../db/connection";

/** Canonical user record. */
export interface User {
  id: number;
  email: string;
  passwordHash: string;
  createdAt: Date;
  lastLogin: Date | null;
  isActive: boolean;
}

/** Insert a new user and return the created record. */
export function createUser(email: string, passwordHash: string): User {
  const conn = getConnection();
  const result = conn.execute(
    "INSERT INTO users (email, password_hash, created_at, is_active) VALUES (?, ?, ?, ?)",
    [email, passwordHash, new Date(), true],
  );
  return {
    id: result.lastId,
    email,
    passwordHash,
    createdAt: new Date(),
    lastLogin: null,
    isActive: true,
  };
}

/** Fetch a user by email, returning a raw record (auth hot path). */
export function getUserByEmail(email: string): User | null {
  const conn = getConnection();
  const { rows } = conn.execute(
    "SELECT id, email, password_hash, created_at, last_login, is_active FROM users WHERE email = ? AND is_active = 1",
    [email],
  );
  if (rows.length === 0) {
    return null;
  }
  return rowToUser(rows[0]);
}

/** Fetch a user by primary key. */
export function getUser(userId: number): User | null {
  const conn = getConnection();
  const { rows } = conn.execute(
    "SELECT id, email, password_hash, created_at, last_login, is_active FROM users WHERE id = ?",
    [userId],
  );
  if (rows.length === 0) {
    return null;
  }
  return rowToUser(rows[0]);
}

/** Stamp the user's lastLogin timestamp. */
export function updateLastLogin(userId: number): void {
  const conn = getConnection();
  conn.execute("UPDATE users SET last_login = ? WHERE id = ?", [new Date(), userId]);
}

/** Soft-delete a user by flipping isActive off. */
export function deactivateUser(userId: number): void {
  const conn = getConnection();
  conn.execute("UPDATE users SET is_active = 0 WHERE id = ?", [userId]);
}

function rowToUser(row: unknown[]): User {
  return {
    id: row[0] as number,
    email: row[1] as string,
    passwordHash: row[2] as string,
    createdAt: row[3] as Date,
    lastLogin: (row[4] as Date | null) ?? null,
    isActive: Boolean(row[5]),
  };
}
